packages/node/src/parseForPolicy.js

```diff
@@ -114,54 +106,100 @@ async function parseForPolicy({
   })
 }
 
-function makeResolveHook({ projectRoot, resolutions = {}, canonicalNameMap }) {
+/**
+ *
+ * @param {string} rootDir
+ * @param {Record<string, string>} canonicalNameMap
+ * @param {Record<string, string>} [resolutions]
+ * @returns {ResolveHook}
+ */
+function makeResolutionResolveHook(rootDir, canonicalNameMap, resolutions) {
+  if (!resolutions || Object.keys(resolutions).length === 0) {
+    // eslint-disable-next-line no-unused-vars
+    return (requestedName, referrer) => null
+  }
   return (requestedName, referrer) => {
     const parentPackageName = getPackageNameForModulePath(
       canonicalNameMap,
       referrer
     )
     // handle resolution overrides
-    const result = checkForResolutionOverride(
-      resolutions,
-      parentPackageName,
-      requestedName
+    return (
+      checkForResolutionOverride(
+        resolutions,
+        parentPackageName,
+        requestedName
+      ) ?? null
+    )
+  }
+}
+
+/**
+ * Creates a resolve hook for a given project root, package.json, and canonical name map.
+ * @param {string} rootDir - The root directory of the project.
+ * @param {Record<string, string>} canonicalNameMap - A map of canonical names to module IDs.
+ * @param {Object} opts - An optional object containing additional options.
+ * @param {Record<string, string>} [opts.resolutions] - A map of package names to paths to override resolution.
+ * @returns {ResolveHook} - A resolve hook function.
+ */
+function makeResolveHook(rootDir, canonicalNameMap, opts = {}) {
+  const getResolutionOverride = makeResolutionResolveHook(
+    rootDir,
+    canonicalNameMap,
+    opts.resolutions
+  )
+
+  return (requestedName, referrer) => {
+    const resolutionOverride = getResolutionOverride(
+      rootDir,
+      requestedName,
+      referrer
     )
-    if (result) {
-      // if path is a relative path, it should be relative to the projectRoot
-      if (path.isAbsolute(result)) {
-        requestedName = result
+
+    if (isBuiltin(requestedName)) {
+      return requestedName
+    }
+
+    requestedName = requestedName || '.'
+
+    if (resolutionOverride) {
+      if (path.isAbsolute(resolutionOverride)) {
+        requestedName = path.relative(rootDir, resolutionOverride)
       } else {
-        requestedName = path.resolve(projectRoot, result)
+        requestedName = resolutionOverride
       }
     }
-    // utilize node's internal resolution algo
-    const { resolve } = createRequire(new URL(`file://${referrer}`))
-    /* eslint-disable no-useless-catch */
-    let resolved
+
+    const { resolve } = Module.createRequire(referrer)
```

Comment:
if you didn't catch it elsewhere: we cannot use the `resolve` package for ESM because it doesn't understand `exports` nor `imports`. it is still used as a fallback elsewhere where we _really_ need the `package.json` and the package doesn't want to give it to us.